core/blockchain.go

```diff
@@ -2071,8 +2071,15 @@ func (bc *BlockChain) reorg(oldBlock, newBlock *types.Block) error {
 	for _, tx := range types.HashDifference(deletedTxs, addedTxs) {
 		rawdb.DeleteTxLookupEntry(indexesBatch, tx)
 	}
-	// Delete any canonical number assignments above the new head
-	number := bc.CurrentBlock().NumberU64()
+	deletedTxs, addedTxs = nil, nil // release the references in case the slices are huge
+
+	// Delete all hash markers that are not part of the new canonical chain.
+	// Because the reorg function does not handle new chain head, all hash
+	// markers greater than or equal to new chain head should be deleted.
+	number := commonBlock.NumberU64()
+	if len(newChain) > 1 {
+		number = newChain[1].NumberU64()
```

Comment:
Because the semantic of `reorg` more like: prepare the chain state for new head block.
I don't plan to refactor or clean up this logic in this PR. Perhaps let's have this tiny fix in first?

Reply:
> This is the same as `bc.CurrentBlock.NumberU64()`, since the loop aove ends on `bc.writeHeadBlock(newChain[1])`, and that method sets `bc.CurrentBlock`.
> 
> So this PR only makes any change if the `len(newChain)` is `0` or `1`. The `0` case should not happen, and for the `1` case, I agree it makes sense to delete everything above the `commonBlock`.

Yes exactly. It mostly for this case `len(newChain)=1`.

Reply:
> Perhaps let's have this tiny fix in first? 

Sounds good to me (just need to get the linter silent)